homeworks/hw1/BWT/Program.cs

```diff
@@ -0,0 +1,110 @@
+            shifts = shifts.Substring(1) + firstChar;
+            shiftTable[i] = shifts;
+        }
+        Array.Sort(shiftTable, StringComparer.Ordinal);
+        string result = "";
+        for (var i = 0; i < shiftTable.Length; i++)
+        {
+            result += shiftTable[i][shiftTable.Length - 1];
+        }
+        return (result, Array.IndexOf(shiftTable, originalString));
+    }
+
+    private const int alphabetSize = 65536;
+
+    // Reverse BWT, takes string after BWT and index of original string in sorted shifts table,
+    // returns original string
+    static string ReverseConversion(string convertedString, int tableIndex)
+    {
+        var countSortArray = new int[alphabetSize];
+        for (var i = 0; i < convertedString.Length; ++i)
+        {
+            ++countSortArray[convertedString[i]];
+        }
+
+        // Sort symbols to get first column of original table
+        //countSortArray[i] to first position i-symbol in 1st column
+        var sum = 0;
+        for (var i = 0; i < alphabetSize; ++i)
+        {
+            sum += countSortArray[i];
+            countSortArray[i] = sum - countSortArray[i];
+        }
+
+        var nextSymbolIndex = new int[convertedString.Length];
+        for (var i = 0; i < nextSymbolIndex.Length; ++i)
+        {
+            nextSymbolIndex[countSortArray[convertedString[i]]] = i;
+            countSortArray[convertedString[i]]++;
+        }
+
+        var nextSymbol = nextSymbolIndex[tableIndex];
+        var answer = new char[convertedString.Length];
+        for (int i = 0; i < convertedString.Length; ++i)
+        {
+            answer[i] = convertedString[nextSymbol];
+            nextSymbol = nextSymbolIndex[nextSymbol];
+        }
+        return new string(answer);
+    }
+
+    static bool Tests()
+    {
+        var result = SlowDirectConversion("ABACABA");
+        if (result.Item1 != "BCABAAA" || result.Item2 != 2)
+        {
+            return false;
+        }
+        result = SlowDirectConversion("baNanaNaS");
+        if (result.Item1 != "aaanbNNSa" || result.Item2 != 7)
+        {
+            return false;
+        }
+        result = SlowDirectConversion("you want a different string instance");
+        if (result.Item1 != "tagtu twn rfcfindr aiiaeyetn nnsso e" || result.Item2 != 35)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    static void Main(string[] args)
+    {
+        if (!Tests())
+        {
+            Console.WriteLine("Tests failed");
+            return;
+        }
+
+        Console.WriteLine("Input a string");
+        string? input = Console.ReadLine();
+        while (input == null)
+        {
+            Console.WriteLine("Incorrect input.\n Waiting for string");
+            input = Console.ReadLine();
+        }
+        var directBWTResult = SlowDirectConversion(input);
+        var stringAfterBWT = directBWTResult.Item1;
+        var inputStringIndex = directBWTResult.Item2;
```

Comment:
```suggestion
        var (stringAfterBWT, inputStringIndex) = SlowDirectConversion(input);
```
C# так тоже умеет :)